src/Sarif.Driver/Sdk/IncompatibleRuleHandling.cs

```diff
@@ -0,0 +1,12 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.CodeAnalysis.Sarif.Driver
+{
+    public enum IncompatibleRuleHandling
```

Comment:
>u [](http://example.com/codeflow?start=29&length=1)

You have implemented these choices cleanly. But can you explain why you think we need them? :)

If someone specifies 'ignore', then why would they bother to publish their incompatible rule ids?

Is it clear to us that a skimmer would have a single setting that would apply to all incompatible rules? 

Why would a rule choose to disable and continue analysis?

What happens if two incompatible rules have different settings? If either rule chooses 'exit analysis' then we will exit. If both rules are disabled, then analysis isn't complete.

This additional machinery would be interesting if it could be specified via configuration. But if a user can specify configuration, they could simply choose to selectively enable/disable the rule they want to run.

I am interested in your thoughts, of course, but at this point I wonder whether we shouldn't simply take the clean 'exit analysis' behavior, which would be a smaller change. We could improve on it later. What's your opinion?

Reply:
The reason I am thinking about adding the `DisableAndContinueAnalysis` is, adding a new rule with existing rule as incompatible rule should not break existing analysis.

For example: The SARIF web validator scenario. If we add the new `GH1007` ProvideFullyFormattedMessageStrings, its incompatible with 'SARIF2002' ProvideMessageArguments. The validator allows users to run all SARIF* rules with GH* rule together. After the new GH1007 rule added, users run the SARIF and GH rules together, they will get a tool configuration exception indicates there is a incompatible rule, they will not see any results. The validator needs to define 2 set of configuration files to enable SARIF2002 and disable GH1007 or the opposite, and switch configuration file between these 2. So I think an option to just disable the incompatible rule and continue analysis may be useful in this case. 
What do you think

Reply:
I think the exception is appropriate. When you add a new GH rule, you will be depending on existing configuration. The existing rule can't reasonably be expected to defensively turn itself off if a future rule marks it as incompatible (and what if you want the new rule to run by default, as we do with the GH analysis??).

Really, we can't handle the full matrix of configuration cases in a reasonable way, I think. Let's remove this extra complexity and just fail-fast in the event of incompatibility. For sure, we need to mark new rules with incompatibilities as a breaking change. 

Please push back if you disagree and/or see deeper into this topic than I. :)

Reply:
ok i agree, there are more complicated cases of handling the conflicting rules, we can start with the throwing exception now.